standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -7040,6 +7040,40 @@ private boolean updatePartitonColStatsInternal(Table tbl, ColumnStatistics colSt
     return parameters != null;
   }
 
+  private boolean updatePartitionColStatsInBatch(Table tbl, Map<String, ColumnStatistics> statsMap,
+                                                 String validWriteIds, long writeId)
+          throws MetaException, InvalidObjectException, NoSuchObjectException, InvalidInputException {
+
+    if (statsMap.size() == 0) {
+      return false;
+    }
+
+    String catalogName = tbl.getCatName();
+    String dbName = tbl.getDbName();
+    String tableName = tbl.getTableName();
+
+    startFunction("updatePartitionColStatsInBatch", ":  db=" + dbName  + " table=" + tableName);
+
+    Map<String, ColumnStatistics> newStatsMap = new HashMap<>();
+    for (Map.Entry entry : statsMap.entrySet()) {
+      ColumnStatistics colStats = (ColumnStatistics) entry.getValue();
+      normalizeColStatsInput(colStats);
+      assert catalogName.equalsIgnoreCase(colStats.getStatsDesc().getCatName());
+      assert dbName.equalsIgnoreCase(colStats.getStatsDesc().getDbName());
+      assert tableName.equalsIgnoreCase(colStats.getStatsDesc().getTableName());
+      newStatsMap.put((String) entry.getKey(), colStats);
+    }
+
+    boolean ret = false;
+    try {
+      ret = getTxnHandler().updatePartitionColumnStatistics(newStatsMap, this,
```

Comment:
Do you need to pass the reference back to HMSHandler inside of TxnHandler? Could this be refactored. This could cause multiple loopholes.

Reply:
done